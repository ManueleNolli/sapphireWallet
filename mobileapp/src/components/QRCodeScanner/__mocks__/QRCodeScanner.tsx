const QRCodeScannerMock = () => (
  <div data-testid={'QRCodeScanner'}>QRCodeScanner</div>
)

export default QRCodeScannerMock
