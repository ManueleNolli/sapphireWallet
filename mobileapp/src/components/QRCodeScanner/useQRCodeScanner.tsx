import { useEffect, useState } from 'react'
import { BarCodeScanner, PermissionStatus } from 'expo-barcode-scanner'

type QRCodeScannerProps = {
  onQRCodeScanned: (data: string) => void
}

export default function useQRCodeScanner({
  onQRCodeScanned,
}: QRCodeScannerProps) {
  const [hasPermission, setHasPermission] = useState<boolean>(false)
  const [scanned, setScanned] = useState(false)

  useEffect(() => {
    const getBarCodeScannerPermissions = async () => {
      const { status } = await BarCodeScanner.requestPermissionsAsync()
      setHasPermission(status === PermissionStatus.GRANTED)
    }
    if (!hasPermission) getBarCodeScannerPermissions()
  }, [])

  const handleBarCodeScanned = ({ data }: { data: string }) => {
    setScanned(true)
    onQRCodeScanned(data)
  }

  return { scanned, handleBarCodeScanned }
}
