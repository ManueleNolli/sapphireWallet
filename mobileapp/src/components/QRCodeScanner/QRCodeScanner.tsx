import React from 'react'
import { Image } from 'react-native'
import { StyleService, useStyleSheet } from '@ui-kitten/components'
import { appStyles, vh, vw } from '../../Styles'
import { BarCodeScanner } from 'expo-barcode-scanner'
import { Platform, View } from 'react-native'
import { qrCodeSmall } from '../../assets/AssetsRegistry'
import useQRCodeScanner from './useQRCodeScanner'

type QRCodeScannerProps = {
  onQRCodeScanned: (data: string) => void
}

export default function QRCodeScanner({ onQRCodeScanned }: QRCodeScannerProps) {
  const { scanned, handleBarCodeScanned } = useQRCodeScanner({
    onQRCodeScanned,
  })
  const styles = useStyleSheet(themedStyles)

  const getCameraStyle = () => {
    // if android = 4:3
    if (Platform.OS === 'android') {
      return {
        height: (4 / 3) * 95 * vw,
        width: 95 * vw,
      }
    }

    return {
      height: 90 * vh,
      width: 95 * vw,
    }
  }

  return (
    <View style={[getCameraStyle(), styles.cameraContainer]}>
      <BarCodeScanner
        onBarCodeScanned={scanned ? undefined : handleBarCodeScanned}
        style={appStyles.center}
      >
        <Image
          style={{
            height: 60 * vw,
            width: 60 * vw,
            opacity: 0.8,
          }}
          source={qrCodeSmall}
        />
      </BarCodeScanner>
    </View>
  )
}

const themedStyles = StyleService.create({
  cameraContainer: {
    borderRadius: 10,
    overflow: 'hidden',
  },
})
