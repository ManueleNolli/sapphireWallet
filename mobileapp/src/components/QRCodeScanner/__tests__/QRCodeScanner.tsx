import React from 'react'
import { act, fireEvent, waitFor } from '@testing-library/react-native'
import renderWithTheme from '../../../TestHelper'
import QRCodeScanner from '../QRCodeScanner'
import useQRCodeScanner from '../useQRCodeScanner'
import { Platform } from 'react-native'

jest.mock('../useQRCodeScanner', () => jest.fn())

describe('QRCodeScanner', () => {
  it('renders correctly not android', () => {
    const onQrCodeScanned = jest.fn()

    ;(useQRCodeScanner as jest.Mock).mockReturnValue({
      scanned: false,
      handleBarCodeScanned: onQrCodeScanned,
    })
    const tree = renderWithTheme(
      <QRCodeScanner onQRCodeScanned={onQrCodeScanned} />
    )

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly android', () => {
    const onQrCodeScanned = jest.fn()
    ;(useQRCodeScanner as jest.Mock).mockReturnValue({
      scanned: false,
      handleBarCodeScanned: onQrCodeScanned,
    })

    Platform.OS = 'android'

    const tree = renderWithTheme(
      <QRCodeScanner onQRCodeScanned={onQrCodeScanned} />
    )

    expect(tree).toMatchSnapshot()
  })
})
