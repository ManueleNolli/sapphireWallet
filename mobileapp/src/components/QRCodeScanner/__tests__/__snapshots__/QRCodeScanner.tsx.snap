// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCodeScanner renders correctly android 1`] = `
<View
  style={
    [
      {
        "height": 949.9999999999999,
        "width": 712.5,
      },
      {
        "borderRadius": 10,
        "overflow": "hidden",
      },
    ]
  }
>
  <ViewManagerAdapter_ExpoBarCodeScanner
    barCodeTypes={[]}
    onBarCodeScanned={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={1}
      style={
        {
          "height": 450,
          "opacity": 0.8,
          "width": 450,
        }
      }
    />
  </ViewManagerAdapter_ExpoBarCodeScanner>
</View>
`;

exports[`QRCodeScanner renders correctly not android 1`] = `
<View
  style={
    [
      {
        "height": 1200.6,
        "width": 712.5,
      },
      {
        "borderRadius": 10,
        "overflow": "hidden",
      },
    ]
  }
>
  <ViewManagerAdapter_ExpoBarCodeScanner
    barCodeTypes={[]}
    onBarCodeScanned={[Function]}
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={1}
      style={
        {
          "height": 450,
          "opacity": 0.8,
          "width": 450,
        }
      }
    />
  </ViewManagerAdapter_ExpoBarCodeScanner>
</View>
`;
